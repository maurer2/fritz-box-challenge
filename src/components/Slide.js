import React, { PureComponent } from 'react';
import styled, { css } from 'styled-components/macro';
import PropTypes from 'prop-types';
import upperFirst from 'lodash/upperFirst';

const Wrapper = styled.div`
  user-select: none;
  text-align: center;
  will-change: transform;
`;

const TitleWrapper = styled.h1`
  margin-top: 0;
  margin-bottom: 2vh;
  color: white;
  font-size: 5vw;
`;

const TextWrapper = styled.h2`
  margin: 0;
  color: white;
  font-size: ${props => 100 / props.characterCount * 1.5}vw;
  line-height: 1;

  /* color: ${props => (props.characterCount === 5 ? 'green' : 'white')}; */
  /*
  ${props => props.numberOfCharacters === 5 && css`
    background: red;
  `};
  */
`;

class TextComponent extends PureComponent {
  get textSeparate() {
    return this.props.text.split('');
  }

  get numberOfCharacters() {
    return this.textSeparate.length;
  }

  render() {
    return (
      <Wrapper>
        <TitleWrapper>
          { upperFirst(this.props.title) }
        </TitleWrapper>
        <TextWrapper characterCount={ this.numberOfCharacters }>
          { this.textSeparate.map(
            (character, index) => (
            <span className="character" key={ index }>
              { character }
            </span>
            ),
          )}
        </TextWrapper>
      </Wrapper>
    );
  }
}

TextComponent.propTypes = {
  text: PropTypes.string,
  title: PropTypes.string,
};

export default TextComponent;
