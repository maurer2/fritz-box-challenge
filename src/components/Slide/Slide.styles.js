import styled from 'styled-components/macro';

export const Wrapper = styled.div`
  user-select: none;
  text-align: center;
  will-change: transform;
`;

export const TitleWrapper = styled.h2`
  margin-top: 0;
  margin-bottom: 2vh;
  font-size: 5vw;
  color: ${({ theme }) => theme.primaryColor};
`;

export const TextWrapper = styled.div`
  margin: 0;
  font-size: ${(props) => Math.floor((100 / props.characterCount) * 1.35)}vw;
  line-height: 1;
  font-weight: bold;
  color: ${({ theme }) => theme.primaryColor};
`;
