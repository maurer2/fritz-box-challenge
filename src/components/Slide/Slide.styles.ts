import styled from 'styled-components';

import type { TextWrapperStyleProps } from './Slide.types';

export const Wrapper = styled.div`
  text-align: center;
  user-select: none;
  will-change: transform;
`;

export const TitleWrapper = styled.h2`
  margin-top: 0;
  margin-bottom: 2vh;
  color: ${({ theme }) => theme.primaryColor};
  font-size: 5vw;
`;

export const TextWrapper = styled.div<TextWrapperStyleProps>`
  margin: 0;
  color: ${({ theme }) => theme.primaryColor};
  font-weight: bold;
  font-size: ${(props) => Math.floor((100 / props.$characterCount) * 1.35)}vw;
  line-height: 1;
`;
