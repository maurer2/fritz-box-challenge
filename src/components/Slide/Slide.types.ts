export interface SlideProps {
  title: string;
  text: string;
}

export interface TextWrapperStyleProps {
  characterCount: number;
}
