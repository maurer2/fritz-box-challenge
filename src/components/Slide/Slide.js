import React from 'react';
import PropTypes from 'prop-types';
import { upperFirst } from 'lodash';

import * as Styles from './Slide.styles';

const Slide = ({ title, text }) => (
  <Styles.Wrapper>
    <Styles.TitleWrapper>
      { upperFirst(title) }
    </Styles.TitleWrapper>
    <Styles.TextWrapper characterCount={text.length}>
      {text.split('').map((character, index) => (
        <span className="character" key={index}>
          { (index === 0) ? upperFirst(character) : character }
        </span>
      ))}
    </Styles.TextWrapper>
  </Styles.Wrapper>
);

const { string } = PropTypes;

Slide.propTypes = {
  text: string.isRequired,
  title: string.isRequired,
};

export { Slide };
