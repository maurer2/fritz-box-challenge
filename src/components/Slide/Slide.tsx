import React from 'react';
import PropTypes from 'prop-types';
import { upperFirst } from 'lodash';

import * as Styles from './Slide.styles';
import * as Types from './Slide.types';

const Slide: React.FC<Types.SlideProps> = ({ title, text }): JSX.Element => (
  <Styles.Wrapper>
    <Styles.TitleWrapper>
      { upperFirst(title) }
    </Styles.TitleWrapper>
    <Styles.TextWrapper characterCount={text.length}>
      {text.split('').map((character, index) => (
        <span className="character" key={index}>
          { (index === 0) ? upperFirst(character) : character }
        </span>
      ))}
    </Styles.TextWrapper>
  </Styles.Wrapper>
);

const { string } = PropTypes;

Slide.propTypes = {
  text: string.isRequired,
  title: string.isRequired,
};

export { Slide };
