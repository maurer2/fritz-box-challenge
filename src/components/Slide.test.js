import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'jest-styled-components';
import toJson from 'enzyme-to-json';

import Slide from './Slide';

// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

describe('Slide', () => {
  const wrapper = shallow(<Slide
    text="123456"
    title="title"
  />);

  const wrapperDeep = mount(<Slide
    text="123456"
    title="title"
  />);

  it('should match snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should should contain spans with numbers', () => {
    expect(wrapper.find('.character').length).toBeGreaterThan(0);
    expect(wrapper.find('.character').length).toBe(6);
  });

  it('should have a wrapper', () => {
    expect(wrapper.find('Wrapper').length).toBe(1);
  });

  it('should have a title', () => {
    expect(wrapper.find('TitleWrapper').length).toBe(1);
    expect(wrapper.find('TitleWrapper').text()).toBe('Title');
  });

  it('should have a text wrapper with props', () => {
    expect(wrapper.find('TextWrapper').length).toBe(1);

    expect(wrapper.find('TextWrapper').props().characterCount).toBe(6);
  });

  it('Wrapper has correct styling', () => {
    expect(wrapperDeep.find('Wrapper')).toHaveStyleRule('user-select', 'none');
  });

  it('TitleWrapper has correct styling', () => {
    expect(wrapperDeep.find('TitleWrapper')).toHaveStyleRule('font-size', '5vw');
  });

  it('TextWrapper has correct styling', () => {
    expect(wrapperDeep.find('TextWrapper')).toHaveStyleRule('font-weight', 'bold');

    wrapperDeep.setProps({ text: 'short' });
    expect(wrapperDeep.find('TextWrapper')).toHaveStyleRule('font-size', '27vw');

    wrapperDeep.setProps({ text: 'verylong' });
    expect(wrapperDeep.find('TextWrapper')).toHaveStyleRule('font-size', '16vw');
  });
});
