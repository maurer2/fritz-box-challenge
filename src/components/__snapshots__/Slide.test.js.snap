// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slide should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextComponent
    text="123456"
    title="title"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <ForwardRef>
          Title
        </ForwardRef>,
        <ForwardRef
          characterCount={6}
        >
          <span
            className="character"
          >
            1
          </span>
          <span
            className="character"
          >
            2
          </span>
          <span
            className="character"
          >
            3
          </span>
          <span
            className="character"
          >
            4
          </span>
          <span
            className="character"
          >
            5
          </span>
          <span
            className="character"
          >
            6
          </span>
        </ForwardRef>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Title",
        },
        "ref": null,
        "rendered": "Title",
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "TitleWrapper-sc-1qe5c9r-1",
            "isStatic": true,
            "rules": Array [
              "margin-top:0;margin-bottom:2vh;color:white;font-size:5vw;",
            ],
          },
          "displayName": "TitleWrapper",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "TitleWrapper-sc-1qe5c9r-1",
          "target": "h1",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "characterCount": 6,
          "children": Array [
            <span
              className="character"
            >
              1
            </span>,
            <span
              className="character"
            >
              2
            </span>,
            <span
              className="character"
            >
              3
            </span>,
            <span
              className="character"
            >
              4
            </span>,
            <span
              className="character"
            >
              5
            </span>,
            <span
              className="character"
            >
              6
            </span>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "host",
            "props": Object {
              "children": "1",
              "className": "character",
            },
            "ref": null,
            "rendered": "1",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "host",
            "props": Object {
              "children": "2",
              "className": "character",
            },
            "ref": null,
            "rendered": "2",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "host",
            "props": Object {
              "children": "3",
              "className": "character",
            },
            "ref": null,
            "rendered": "3",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": "3",
            "nodeType": "host",
            "props": Object {
              "children": "4",
              "className": "character",
            },
            "ref": null,
            "rendered": "4",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": "4",
            "nodeType": "host",
            "props": Object {
              "children": "5",
              "className": "character",
            },
            "ref": null,
            "rendered": "5",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": "5",
            "nodeType": "host",
            "props": Object {
              "children": "6",
              "className": "character",
            },
            "ref": null,
            "rendered": "6",
            "type": "span",
          },
        ],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "TextWrapper-sc-1qe5c9r-2",
            "isStatic": false,
            "rules": Array [
              "margin:0;color:white;font-size:",
              [Function],
              "vw;line-height:1;",
            ],
          },
          "displayName": "TextWrapper",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "TextWrapper-sc-1qe5c9r-2",
          "target": "h2",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "Wrapper-sc-1qe5c9r-0",
        "isStatic": true,
        "rules": Array [
          "user-select:none;text-align:center;will-change:transform;",
        ],
      },
      "displayName": "Wrapper",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "Wrapper-sc-1qe5c9r-0",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <ForwardRef>
            Title
          </ForwardRef>,
          <ForwardRef
            characterCount={6}
          >
            <span
              className="character"
            >
              1
            </span>
            <span
              className="character"
            >
              2
            </span>
            <span
              className="character"
            >
              3
            </span>
            <span
              className="character"
            >
              4
            </span>
            <span
              className="character"
            >
              5
            </span>
            <span
              className="character"
            >
              6
            </span>
          </ForwardRef>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Title",
          },
          "ref": null,
          "rendered": "Title",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "TitleWrapper-sc-1qe5c9r-1",
              "isStatic": true,
              "rules": Array [
                "margin-top:0;margin-bottom:2vh;color:white;font-size:5vw;",
              ],
            },
            "displayName": "TitleWrapper",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "TitleWrapper-sc-1qe5c9r-1",
            "target": "h1",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "characterCount": 6,
            "children": Array [
              <span
                className="character"
              >
                1
              </span>,
              <span
                className="character"
              >
                2
              </span>,
              <span
                className="character"
              >
                3
              </span>,
              <span
                className="character"
              >
                4
              </span>,
              <span
                className="character"
              >
                5
              </span>,
              <span
                className="character"
              >
                6
              </span>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "host",
              "props": Object {
                "children": "1",
                "className": "character",
              },
              "ref": null,
              "rendered": "1",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "host",
              "props": Object {
                "children": "2",
                "className": "character",
              },
              "ref": null,
              "rendered": "2",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "host",
              "props": Object {
                "children": "3",
                "className": "character",
              },
              "ref": null,
              "rendered": "3",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": "3",
              "nodeType": "host",
              "props": Object {
                "children": "4",
                "className": "character",
              },
              "ref": null,
              "rendered": "4",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": "4",
              "nodeType": "host",
              "props": Object {
                "children": "5",
                "className": "character",
              },
              "ref": null,
              "rendered": "5",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": "5",
              "nodeType": "host",
              "props": Object {
                "children": "6",
                "className": "character",
              },
              "ref": null,
              "rendered": "6",
              "type": "span",
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "TextWrapper-sc-1qe5c9r-2",
              "isStatic": false,
              "rules": Array [
                "margin:0;color:white;font-size:",
                [Function],
                "vw;line-height:1;",
              ],
            },
            "displayName": "TextWrapper",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "TextWrapper-sc-1qe5c9r-2",
            "target": "h2",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "Wrapper-sc-1qe5c9r-0",
          "isStatic": true,
          "rules": Array [
            "user-select:none;text-align:center;will-change:transform;",
          ],
        },
        "displayName": "Wrapper",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "Wrapper-sc-1qe5c9r-0",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
